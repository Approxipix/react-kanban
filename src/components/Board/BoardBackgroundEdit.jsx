import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { editBoardColor } from '../../redux/boardReducer/actions/actions';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import styled from 'styled-components'

const Container = styled.div`
  position: relative;
`;

const List = styled.ul`
  position: absolute;
  top: 90%;
  display: grid;
  grid-template-columns: repeat(2, 2fr);
  grid-gap: .75rem;
  width: 100%;
`;

const Item = styled.li`
`;

const Button = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 5rem;
  background-color: ${props => props.value};
  border-radius: .5rem;
  color: #fff;
  cursor: pointer;
  &:hover {
    box-shadow: inset 0 0 0 10rem rgba(0, 0, 0, .1)
  }
`;

class BoardBackgroundEdit extends Component {
  editBoardColor = (color) => {
    this.props.actions.editBoardColor({
      boardId: this.props.boardId,
      boardColor: color
    });
  };

  render() {
    const { boardColor, colors } = this.props;
    return (
      <Container>
        <List>
          {colors.map((color, index) => (
            <Item key={index}>
              <Button value={color} onClick={() => this.editBoardColor(color)}>
                {boardColor === color && <FontAwesomeIcon icon="check" />}
              </Button>
            </Item>
          ))}
        </List>
      </Container>
    );
  }
}

function mapStateToProps(state) {
  return {
    colors: state.rootReducer.colors,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editBoardColor: editBoardColor,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(BoardBackgroundEdit);
