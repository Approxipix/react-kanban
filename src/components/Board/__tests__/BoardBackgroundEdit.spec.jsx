import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import BoardBackgroundEdit from '../BoardBackgroundEdit';

const mockStore = configureStore([]);

describe('<BoardBackgroundEdit>', () => {
  let store;
  let component;
  let mockToggleSidebar = jest.fn();

  beforeEach(() => {
    const props = {
      boardId: 'BoardID',
      boardColor: ''
    };
    store = mockStore({
      rootReducer: {
        colors: ['#2E7EAF', '#00603d', '#D29034'],
      }
    });
    store.dispatch = mockToggleSidebar;

    const wrapper = shallow(<BoardBackgroundEdit store={store} {...props} />);
    component = wrapper.find('BoardBackgroundEdit').dive();
  });

  it('should renders without crashing', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render a item for each colors', () => {
    expect(component.find('List').children().length).toEqual(3)
  });

  it('should sets the color prop as the `value` on the item', () => {
    const list = component.find('List');

    expect(list.childAt(0).find('Button').props().value).toEqual('#2E7EAF');
    expect(list.childAt(1).find('Button').props().value).toEqual('#00603d');
    expect(list.childAt(2).find('Button').props().value).toEqual('#D29034');
  });

  it("should dispatch an action after click on item", () => {
    const expectedAction = {
      type: 'EDIT_BOARD_COLOR',
      payload: {
        boardId: 'BoardID',
        boardColor: '#2E7EAF'
      }
    };

    component.find('List').childAt(0).find('Button').simulate('click');

    expect(mockToggleSidebar).toHaveBeenCalledTimes(1);
    expect(mockToggleSidebar).toHaveBeenCalledWith(expectedAction);
  });

  it('should render fontawesome icon if boardColor and color are the same', () => {
    const props = {
      boardId: 'BoardID',
      boardColor: '#2E7EAF'
    };
    store = mockStore({
      rootReducer: {
        colors: ['#2E7EAF', '#00603d', '#D29034'],
      }
    });
    const wrapper = shallow(<BoardBackgroundEdit store={store} {...props} />);
    const component = wrapper.find('BoardBackgroundEdit').dive();

    expect(toJson(component)).toMatchSnapshot();
  });
});

