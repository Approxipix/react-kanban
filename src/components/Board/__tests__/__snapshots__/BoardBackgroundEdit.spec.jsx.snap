// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BoardBackgroundEdit> should render fontawesome icon if boardColor and color are the same 1`] = `
<styled.div>
  <List>
    <styled.li
      key="0"
    >
      <Button
        onClick={[Function]}
        value="#2E7EAF"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon="check"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </Button>
    </styled.li>
    <styled.li
      key="1"
    >
      <Button
        onClick={[Function]}
        value="#00603d"
      />
    </styled.li>
    <styled.li
      key="2"
    >
      <Button
        onClick={[Function]}
        value="#D29034"
      />
    </styled.li>
  </List>
</styled.div>
`;

exports[`<BoardBackgroundEdit> should renders without crashing 1`] = `
<styled.div>
  <List>
    <styled.li
      key="0"
    >
      <Button
        onClick={[Function]}
        value="#2E7EAF"
      />
    </styled.li>
    <styled.li
      key="1"
    >
      <Button
        onClick={[Function]}
        value="#00603d"
      />
    </styled.li>
    <styled.li
      key="2"
    >
      <Button
        onClick={[Function]}
        value="#D29034"
      />
    </styled.li>
  </List>
</styled.div>
`;
